File: android/src/main/java/com/bil/bilmobileads/flutter/bilmobileadsflutter/FADInterstitial.java
package com.bil.bilmobileads.flutter.bilmobileadsflutter;

import android.util.Log;

import com.bil.bilmobileads.ADInterstitial;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class FADInterstitial implements MethodChannel.MethodCallHandler {

    public static final FADInterstitial instance = new FADInterstitial();

    private FlutterPlugin.FlutterPluginBinding binding;
    private MethodChannel channel;

    private Map<Integer, ADInterstitial> allAds;
    private Map<Integer, MethodChannel> allAdsDelegate;

    public void register(FlutterPlugin.FlutterPluginBinding pluginBinding) {
        channel = new MethodChannel(pluginBinding.getBinaryMessenger(), Utils.INTERSTITIAL);
        channel.setMethodCallHandler(this);

        allAds = new HashMap<>();
        allAdsDelegate = new HashMap<>();

        binding = pluginBinding;
    }

    public void dispose() {
        channel.setMethodCallHandler(null);
        channel = null;
        binding = null;
        allAds.clear();
        allAds = null;
        allAdsDelegate.clear();
        allAdsDelegate = null;
    }

    public ADInterstitial isPluginReady(int id, String callMethod) {
        ADInterstitial adObj = allAds.get(id);
        if (adObj == null) {
            if (!callMethod.equals("create"))
                Log.d(Utils.LOGTAG, "ADInterstitial " + id + " is null. You need init ADInterstitial first.");
            return null;
        }
        return adObj;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        if (call.method.equals("create") && !FPBMobileAds.instance.isPluginReady()) return;

        int id = call.argument("id");
        ADInterstitial interstitial = isPluginReady(id, call.method);

        if (interstitial != null)
            Log.d(Utils.LOGTAG, "Call - Func: " + call.method + (call.arguments != null ? " | args: " + call.arguments : ""));
        switch (call.method) {
            case "setListener":
                if (interstitial == null) result.error(null, null, null);

                if (interstitial.adDelegate != null) return;

                MethodChannel adChannel = new MethodChannel(binding.getBinaryMessenger(), Utils.INTERSTITIAL + "_" + id);
                allAdsDelegate.put(id, adChannel);
                interstitial.setListener(Utils.createAdListener(adChannel));

                result.success(null);
                break;
            case "create":
                if (interstitial == null) {
                    String adUnitId = (String) call.argument("adUnitId");
                    ADInterstitial adinterstitial = new ADInterstitial(adUnitId);
                    allAds.put(id, adinterstitial);
                }

                result.success(null);
                break;
            case "preLoad":
                if (interstitial == null) result.error(null, null, null);
                interstitial.preLoad();

                break;
            case "show":
                if (interstitial == null) result.error(null, null, null);
                interstitial.show();

                break;
            case "isReady":
                if (interstitial == null) result.error(null, null, null);

                if (interstitial.isReady()) result.success(true);
                else result.success(false);

                break;
            case "destroy":
                if (interstitial == null) result.error(null, null, null);

                interstitial.destroy();
                allAds.remove(id);
                if (allAdsDelegate.get(id) != null) allAdsDelegate.remove(id);

                break;
            default:
                Log.d(Utils.LOGTAG, "Ignoring invoke from native. This normally shouldn\'t happen.");
                result.notImplemented();
                break;
        }
    }
}

File: ios/Classes/FADInterstitial.swift
//
//  FADInterstitial.swift
//  bilmobileadsflutter
//
//  Created by HNL_MAC on 12/28/20.
//

import Flutter
import BilMobileAds

public class FADInterstitial: NSObject, FlutterPlugin {
    
    fileprivate var allAds: [Int: ADInterstitial] = [:]
    fileprivate var adDelegates: [Int: FADIntersitialDelegate] = [:]
    fileprivate var pluginRegistrar: FlutterPluginRegistrar?
    
    fileprivate var adUnitId: String?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FADInterstitial()
        instance.pluginRegistrar = registrar
        let channel = FlutterMethodChannel(name: Utils.INTERSTITIAL, binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    private func isPluginReady(id: Int, callMethod: String) -> ADInterstitial? {
        guard let adObj: ADInterstitial = allAds[id] else {
            if callMethod != "create" {
                print("\(Utils.LOGTAG) -> ADInterstitial \(id) is nil. You need init ADInterstitial first.")
                return nil
            }
            return nil
        }
        return adObj
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if (call.method == "create" && !FPBMobileAds.shared.isPluginReady()) { return }
        
        guard let args = call.arguments as? [String : Any]  else {
            result(FlutterError(code: "Missing args!", message: "Unable to convert args to [String : Any]", details: nil))
            return
        }
        
        let id = args["id"] as! Int
        let interstitial = self.isPluginReady(id: id, callMethod: call.method);
        if interstitial != nil {
            print("\(Utils.LOGTAG) -> Call - Func: \(call.method) | args: \(String(describing: call.arguments != nil ? call.arguments : "")) ");
        }
        switch call.method {
        case "setListener":
            if interstitial == nil { return }

            let channel = FlutterMethodChannel(name: "\(Utils.INTERSTITIAL)_\(id)", binaryMessenger: pluginRegistrar!.messenger())
            let adDelegate = FADIntersitialDelegate(channel: channel)
            interstitial!.setListener(adDelegate)
            adDelegates[id] = adDelegate
            break
        case "create":
            if (interstitial == nil) {
                let adUnitId = args["adUnitId"] as! String
                let adInterstitial = ADInterstitial(FPBMobileAds.shared.getUIViewController(), placement: adUnitId)
                allAds[id] = adInterstitial
            }
            result(nil)
            break
        case "preLoad":
            if interstitial == nil { return }
            interstitial!.preLoad()
            break
        case "show":
            if interstitial == nil { return }
            interstitial!.show()
            break
        case "isReady":
            if interstitial == nil { return }
            result(interstitial!.isReady())
            break
        case "destroy":
            if interstitial == nil { return }
            
            interstitial?.destroy()
            allAds.removeValue(forKey: id)
            adDelegates.removeValue(forKey: id)
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

class FADIntersitialDelegate: NSObject, ADInterstitialDelegate {
    let channel: FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    func interstitialDidReceiveAd() {
        channel.invokeMethod("loaded", arguments: nil)
    }
    
    func interstitialLoadFail(error: String) {
        channel.invokeMethod("failedToLoad", arguments: error)
    }
    
    func interstitialWillPresentScreen() {
        channel.invokeMethod("opened", arguments: nil)
    }
    
    func interstitialDidFailToPresentScreen() {
        channel.invokeMethod("failedToShow", arguments: nil)
    }
    
    func interstitialWillDismissScreen() {}
    
    func interstitialDidDismissScreen() {
        channel.invokeMethod("closed", arguments: nil)
    }
    
    func interstitialWillLeaveApplication() {
        channel.invokeMethod("clicked", arguments: nil)
        channel.invokeMethod("leftApplication", arguments: nil)
    }
}
